shared/src/balancer_event_handler.rs

```diff
@@ -7,31 +7,94 @@ use crate::{
 use anyhow::{anyhow, Context, Result};
 use contracts::{
     balancer_v2_vault::{
-        self, event_data::PoolRegistered as ContractPoolRegistered, Event as ContractEvent,
+        self,
+        event_data::{
+            PoolRegistered as ContractPoolRegistered, TokensRegistered as ContractTokensRegistered,
+        },
+        Event as ContractEvent,
     },
     BalancerV2Vault,
 };
 use ethcontract::common::DeploymentInformation;
 use ethcontract::{dyns::DynWeb3, Event as EthContractEvent, EventMetadata, H160, H256};
-use std::fmt::Debug;
-use std::ops::RangeInclusive;
+use std::convert::TryInto;
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::Debug,
+    ops::RangeInclusive,
+};
 use tokio::sync::Mutex;
 
-#[derive(Debug)]
+#[derive(Clone, Debug)]
 pub enum BalancerEvent {
     PoolRegistered(PoolRegistered),
+    TokensRegistered(TokensRegistered),
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone, PartialEq)]
 pub struct PoolRegistered {
     pub pool_id: H256,
     pub pool_address: H160,
     pub specialization: PoolSpecialization,
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct TokensRegistered {
+    pub pool_id: H256,
+    pub tokens: Vec<H160>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct WeightedPool {
+    pool_id: H256,
+    pool_address: H160,
+    tokens: Vec<H160>,
+    specialization: PoolSpecialization,
+    block_created: u64,
+}
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct WeightedPoolBuilder {
+    pool_registration: Option<PoolRegistered>,
+    tokens_registration: Option<TokensRegistered>,
+    /// Both Pool and Tokens Registered events are emitted in the same transaction
+    /// https://github.com/balancer-labs/balancer-v2-monorepo/blob/70843e6a61ad11208c1cfabf5cfe15be216ca8d3/pkg/pool-utils/contracts/BasePool.sol#L128-L130
+    /// block_number is only contained in the EventIndex
+    block_created: u64,
+}
+
+impl TryInto<WeightedPool> for WeightedPoolBuilder {
```

Comment:
@vkgnosis convinced me that while these traits are nice (and a sign one knows rust) they are not always practical especially if the object is not passed around publicly (and therefore it's unlikely an external consumer wants to use the TryInto<WeightedPool> abstraction somewhere).

Here I think an `Option<WeightedPool>` would be a slightly more normal return type (but of course Result with void error works as well).

Reply:
Right, so I would be happy to return `Option<WeightedPool>` and it makes perfect sense, but then I suppose this falls outside the scope of `TryInto`, perhaps this is what `Into` is for.

Reply:
The idea is that these pre-existing traits (Into, TryInto, etc) only really make sense when you pass your object to external consumers which may be operating on the more abstract trait type. In other cases it's not bad to use them, but there is no need to shoehorn something that is different into fitting a pre-existing trait.

E.g. a method that is agnostic to weither its using a PoolBuilder or a Pool or something else that can be turned into a Pool, might look like: `foo(impl Into<WeightedPool>)`. This method would benefit from`impl TryInto<WeightedPool> for WeightedPoolBuilder` because it could use both types. 

But here, you are not passing anything around so you might as well just implement `my_try_into` on `Impl WeightedPoolBuilder` directly. I hope that makes sense.